### PR TITLE
[policy] switch to sha256 sum

### PR DESCRIPTION
**Pull request by BryanQuigley**
WIP - need cleaner/sosnode changes

Resolves: #2351
Closes: #1204

Signed-off-by: Bryan Quigley <code@bryanquigley.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X] If this commit closes an existing issue, is the line `Closes: #ISSUENUMBER` included in an independent line?
- [X] If this commit resolves an existing pull request, is the line `Resolves: #PRNUMBER` included in an independent line?

### CODE REVIEW

**Comment by TurboTurtle** [comment]
Thus far I haven't found anything internal on the RH side that would be adversely affected by this change. I'll update again if that changes, but so far this is looking like it is/will be an easy low-impact change.

**Comment by BryanQuigley** [comment]
Haven't tested the collector changes yet

**Comment by TurboTurtle** [comment]
I ran a quick test on the `collect` bits, couple notes:

- There is an attempt to grab the `.sha256` file from the master node, regardless of whether it exists or not. This is likely an existing bug just getting highlighted by the fact that the master in my case was not running this branch. 
- The `.md5` sums that are generated by current releases of sos are not gathered. For this change we'd need `collect` to gather both `.md5` and `.sha256` files if they exist, to account for mix-version node lists.

**Comment by BryanQuigley** [comment]
I found it easier to just switch to the hash outputed via stdout.  In this iteration the checksum just gets added to the json manifest.    Thoughts?

**Comment by TurboTurtle** [comment]
Hmm, I hadn't considered that, but it is a nice alternative to collecting the actual checksum files.

I don't _think_ there'd be any issues with dropping the file collection, and it'd arguably make things easier on any automated inspection of sos-collect tarballs.

I'll find some time this week to run this updated branch and in the meantime see if I can find any "gotcha"s. If not, then I think writing it to the manifest instead is a nice improvement.

**Comment by TurboTurtle** [comment]
Hmmm. Ran this and I'm not seeing a checksum added to the `collect` manifest.

**Comment by BryanQuigley** [comment]
It's certainly possible I missed some bits.. my test case was to a VM running an older version and the localhost:
bin/sos collect --nodes 192.168.122.183 --ssh-user bryan --password

That results in the local collection using from git so sha256sum, and the remote one using md5

**Comment by TurboTurtle** [comment]
Yeah, I'd expect there to be different checksum types for older versions based on the commit, but what I mean is neither are getting written to the `manifest.json` created for the `collect` run.

```
$ sudo ./bin/sos collect --master=redacted.redhat.com --password

[...]
The following archive has been created. Please provide it to your support team.
	/var/tmp/sos-collector-123456-2021-01-29-ppbbu.tar.xz

[.. extract it ..]

[jake@terra sostmp]$ grep md5 sos-collector-123456-2021-01-29-ppbbu/sos_reports/manifest.json 
[jake@terra sostmp]$ grep sha sos-collector-123456-2021-01-29-ppbbu/sos_reports/manifest.json 
[jake@terra sostmp]$ 
```

**Comment by BryanQuigley** [comment]
Hmm, I switched mine to do master node with type none and still find the md5 present. What cluster type are you using?

**Comment by TurboTurtle** [comment]
It's a RHV (oVirt) cluster, however that shouldn't be making a difference since the manifest is handled outside of any cluster profiles.

Just tried again using `--cluster-type none` and I get the same result. No md5/sha sum added to the manifest generated for `collect`.

**Comment by TurboTurtle** [comment]
Just to make sure it wasn't something with my specific workstation, I loaded up a fresh RHEL 8 VM and tried it there as well - same results.

**Comment by BryanQuigley** [comment]
I just did a bunch of attempts including:
One was from my debian desktop to two centos 8 VMs. (bin/sos collect --no-local --nodes 192.168.122.233 --nodes 192.168.122.25 --cluster-type none --password)
One was from a CentOS 8 VM to include another centos VM (bin/sos collect --nodes 192.168.122.233 --cluster-type none --password)

All have md5s added to the manifest.json. Just to confirm your new VM test was just a single node with no cluster setup on it?

**Comment by TurboTurtle** [comment]
Correct, new VM was just a single node.

Out of curiosity can you post a snippet of the manifest with the checksum?

**Comment by BryanQuigley** [comment]
{
    "version": "4.0",
    "cmdline": "collect --no-local --nodes 192.168.122.233 --nodes 192.168.122.25 --cluster-type none --password",
    "start_time": "2021-02-05 10:13:24.422109",
    "end_time": "2021-02-05 10:14:21.263212",
    "run_time": "0:00:56.841103",
    "compression": "auto",
    "policy": "Debian",
    "components": {
        "collect": {
            "cluster_type": null,
            "node_list": [
                "192.168.122.233",
                "192.168.122.25"
            ],
            "initial_sos_cmd": "sosreport --batch -c auto",
            "master": "localhost",
            "nodes": {
                "localhost": {
                    "hostname": "desktop",
                    "policy": "Debian",
                    "sos_version": null,
                    "final_sos_command": ""
                },
                "192.168.122.25": {
                    "hostname": "192.168.122.25",
                    "policy": "CentOS",
                    "sos_version": "3.9.1",
                    "final_sos_command": "/usr/sbin/sosreport --batch -c auto ",
                    "label": "",
                    "md5": "262cdc11fae401a47e982072b1fe660f",
                    "collected_archive": "sosreport-localhost-2021-02-05-hzlnftl.tar.xz"
                },
                "192.168.122.233": {
                    "hostname": "192.168.122.233",
                    "policy": "CentOS",
                    "sos_version": "3.9.1",
                    "final_sos_command": "/usr/sbin/sosreport --batch -c auto ",
                    "label": "",
                    "md5": "8cf67fe3ab0f28b03c6bb7c54b1409a8",
                    "collected_archive": "sosreport-localhost-2021-02-05-mbatihh.tar.xz"
                }
            }
        }
    }
}

**Comment by TurboTurtle** [comment]
Ok, so it's not an ordering issue or anything - on my RHEL family boxes we're just straight up missing the checksum keys, yet the `collected_archive` which is added afterwards is present as expected.

About the only other thing I can think of right now is python version - what version is on the system you're running collect from?

File: sos/collector/sosnode.py [hunk 1]
```diff
@@ -781,6 +781,15 @@ def execute_sos_command(self):
...
                         path = line.strip()
+                    if line.startswith(" sha256\t"):
+                        self.manifest.add_field('checksum', line.split("\t")[1])
+                        self.manifest.add_field('checksum_type', 'sha256')
+                    if line.startswith(" md5\t"):
+                        self.manifest.add_field('checksum', line.split("\t")[1])
+                        self.manifest.add_field('checksum_type', 'md5')
+                    if line.startswith("The checksum is: "):
+                        self.manifest.add_field('checksum', line.split()[3])
+                        self.manifest.add_field('checksum_type', 'unspecified_guess_by_size')
```
Comment (TurboTurtle) [inline sos/collector/sosnode.py:792]: This should be automatically set, as the hash sizes are well-known. E.G. something like:

```python
if line.startswith((" sha256\t", " md5\t")):
    checksum = line.split("\t")[1]
if line.startswith("The checksum is: "):
    checksum = line.split()[3]
self.manifest.add_field('checksum', checksum)
self.manifest.add_field('checksum_type', 'md5' if len(checksum) == 32 else 'sha256')
```

**Review by TurboTurtle — APPROVED** [review]
Ack, tested against ovirt and a random list of nodes.

**Pull request closed without merging**